gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -112,10 +112,26 @@ def self._check_final_ancestors(target, target_ancestors, source_method_names)
             ancestor.private_method_defined?(method_name) ||
             ancestor.protected_method_defined?(method_name)) &&
            final_method?(method_owner_and_name_to_key(ancestor, method_name))
-          raise(
-            "The method `#{method_name}` on #{ancestor} was declared as final and cannot be " +
-            (target == ancestor ? "redefined" : "overridden in #{target}")
-          )
+          definition_file, definition_line = T::Private::Methods.signature_for_method(ancestor.instance_method(method_name)).method.source_location
+          is_redefined = target == ancestor
+          caller_loc = caller_locations&.find {|l| !l.to_s.match?(%r{sorbet-runtime[^/]*/lib/}) }
+          extra_info = "\n"
+          if caller_loc
+            extra_info = (is_redefined ? "Redefined" : "Overridden") + " here: #{caller_loc.path}:#{caller_loc.lineno}\n"
+          end
+
+          error_message = "The method `#{method_name}` on #{ancestor} was declared as final and cannot be " +
+                          (is_redefined ? "redefined" : "overridden in #{target}")
+          pretty_message = "#{error_message}\n" \
+                           "Made final here: #{definition_file}:#{definition_line}\n" \
+                           "#{extra_info}"
+
+          begin
+            raise pretty_message
+          rescue => e
+            T::Private::DeclState.current.reset!
```

Comment:
What's this `DeclState.current.reset!` for?

Reply:
Otherwise it looks like we're still building a sig and we get spurious test failures.

Reply:
from offline: when we were unconditionally raising, it didn't matter whether whether we clean up to a known good state because the `raise` would crash the program anyways. but when the validation handler might not raise, we have to go out of our way to make sure we're setting the state good again.

@froydnj one question: does calling `reset!` here make this read value stale?

https://github.com/sorbet/sorbet/blob/d570e3bac6ccbc5e7678eeef00e8b7763b4a78f8/gems/sorbet-runtime/lib/types/private/methods/_methods.rb#L147

i.e., maybe we should re-get current_declaration after calling `_check_final_ancestors`? not sure what the current behavior is if we're not doing that.

Reply:
To expand on that explanation: we currently always `raise` for `final` errors.  This PR changes things so that we go through `sig_validation_error_handler` for `final` errors and also adds a test that:

1. We do the "correct" thing when `sig_validation_error_handler` is using its default behavior.
2. Things continue to work even if `sig_validation_error_handler` doesn't `raise`.

But this second test causes a problem: after we don't `raise` and return normally, `DeclState` still thinks it's building a sig.  So, depending on which test executes next, we might get errors about attempting to build a sig while we're building a sig.  Therefore, we need this reset to ensure that `DeclState` is consistent no matter what the handler does.

Reply:
(were you planning to leave that comment here, or check it into the code?)

Reply:
I checked in a cut-down version of the comment.